### PR TITLE
Concretizer: put build->run deps in parent unification set (fixes #46034)

### LINKED ISSUE #46034: Duplicate package in PYTHONPATH
### Steps to reproduce

In some cases there are duplicate packages in the `$PYTHONPATH`. We encountered the problem in case of `py-setuptools`.

For a minimal example, we created a new package `test-duplicate`:
```
class TestDuplicate(Package):

    homepage = "http://www.example.com"
    url = "http://www.example.com/pip-1.0.tar.gz"

    version("1", md5="0123456789abcdef0123456789abcdef")

    depends_on('py-setuptools@61:', type='build')
    depends_on("py-jupyter-packaging@0.11.1", type='build')
```

When creating the `build-env`:
```
spack build-env test-duplicate
```

there are two versions of `py-setuptools` in the `$PYTHONPATH`: `PYTHONPATH=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-jupyter-packaging-0.11.1-37tiw4alqmtcngo6gfi2nhtkborpsvhu/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-tomlkit-0.12.1-ruefhdh7k75ludnzo67bc4mh4tacj5ab/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-deprecation-2.1.0-cq5jpqaljxwjvvosbcwpmkbhq576wf3z/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-setuptools-59.4.0-jx7houhznbncur4zbzoon23kdphm3dxa/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-setuptools-69.2.0-mjruwxhsgt3pvqv4vkrqzrfkmawtgy3n/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-packaging-23.1-h6fjzyvoh47jl4asdrfryzdiachbqa3b/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-wheel-0.41.2-uryi3djcgi7stoc5s6g5kj2jxc7w5otl/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/python-venv-1.0-bkja3lbk6zg3o3l2s4skhmz3x57nqgoe/lib/python3.8/site-packages
`

We assume that the problem arises when you have a build env for a package that has:
* a direct build dep on `py-setuptools@A`
* a direct build dep on a package with a run-time dep on `py-setuptools@B` (which in terms of `$PYTHONPATH` translates to a build dep on `py-setuptools@B` for the root package)

While we did not find a issue which directly mentions this bug, I think that the build in #44967 might suffered from the same issue (and was indirectly hidden by declaring `py-setuptools` as a run dependency).

@elenimath @muffgaga 

### Error message

```
SPACK_ROOT=/home/jakob/spack
PATH=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-wheel-0.41.2-uryi3djcgi7stoc5s6g5kj2jxc7w5otl/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/python-venv-1.0-bkja3lbk6zg3o3l2s4skhmz3x57nqgoe/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/python-3.8.2-7ktojpvncdkaoqvvkawwwmfgbdrxxndi/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/tar-1.34-j5z5yr3k4t2fsjwwuq4fujklaa7h4l3a/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/zstd-1.5.6-h6uwy3hfwfmkuobniekmq6zutjtezkme/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/pigz-2.8-wrrgxt2kwa7d2iquxmkdh2qbskznza5l/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/xz-5.4.6-xlreg7q7s4dtc5l2z5b5k6nbda7cppdf/bin:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/bzip2-1.0.8-n2sbebrebf5k7cd6hnjyik4xm5xlgb7r/bin:/home/jakob/spack/lib/spack/env/gcc:/home/jakob/spack/lib/spack/env/case-insensitive:/home/jakob/spack/lib/spack/env:/home/jakob/spack/bin:/home/jakob/.local/bin:/home/jakob/.local/miniconda3/bin:/home/jakob/.local/miniconda3/condabin:/home/jakob/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/go/bin:/usr/local/go/bin
CC=/home/jakob/spack/lib/spack/env/gcc/gcc
CCACHE_DISABLE=1
CMAKE_PREFIX_PATH=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-setuptools-69.2.0-mjruwxhsgt3pvqv4vkrqzrfkmawtgy3n:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-jupyter-packaging-0.11.1-37tiw4alqmtcngo6gfi2nhtkborpsvhu:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/gcc-runtime-12.2.0-u3t2khdtejyh4r7bsmqh67d7vxlq6l7g
CXX=/home/jakob/spack/lib/spack/env/gcc/g++
F77=/home/jakob/spack/lib/spack/env/gcc/gfortran
FC=/home/jakob/spack/lib/spack/env/gcc/gfortran
PYTHONPATH=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-jupyter-packaging-0.11.1-37tiw4alqmtcngo6gfi2nhtkborpsvhu/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-tomlkit-0.12.1-ruefhdh7k75ludnzo67bc4mh4tacj5ab/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-deprecation-2.1.0-cq5jpqaljxwjvvosbcwpmkbhq576wf3z/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-setuptools-59.4.0-jx7houhznbncur4zbzoon23kdphm3dxa/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-setuptools-69.2.0-mjruwxhsgt3pvqv4vkrqzrfkmawtgy3n/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-packaging-23.1-h6fjzyvoh47jl4asdrfryzdiachbqa3b/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/py-wheel-0.41.2-uryi3djcgi7stoc5s6g5kj2jxc7w5otl/lib/python3.8/site-packages:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/python-venv-1.0-bkja3lbk6zg3o3l2s4skhmz3x57nqgoe/lib/python3.8/site-packages
SPACK_CC=/usr/bin/gcc
SPACK_CC_RPATH_ARG=-Wl,-rpath,
SPACK_COMPILER_FLAGS_KEEP=
SPACK_COMPILER_FLAGS_REPLACE=-Werror-|-Wno-error= -Werror|-Wno-error
SPACK_COMPILER_SPEC=gcc@=12.2.0
SPACK_CXX=/usr/bin/g++
SPACK_CXX_RPATH_ARG=-Wl,-rpath,
SPACK_DEBUG_LOG_DIR=/tmp/jakob/spack-stage/spack-stage-test-duplicate-1-yc7izcsh6n757u43vwzccvocqxospehy
SPACK_DEBUG_LOG_ID=test-duplicate-yc7izcs
SPACK_DTAGS_TO_ADD=--disable-new-dtags
SPACK_DTAGS_TO_STRIP=--enable-new-dtags
SPACK_ENV_PATH=/home/jakob/spack/lib/spack/env:/home/jakob/spack/lib/spack/env/case-insensitive:/home/jakob/spack/lib/spack/env/gcc
SPACK_F77=/usr/bin/gfortran
SPACK_F77_RPATH_ARG=-Wl,-rpath,
SPACK_FC=/usr/bin/gfortran
SPACK_FC_RPATH_ARG=-Wl,-rpath,
SPACK_INCLUDE_DIRS=
SPACK_LINKER_ARG=-Wl,
SPACK_LINK_DIRS=
SPACK_MANAGED_DIRS="/home/jakob/spack/opt/spack/"*|"/tmp/jakob/spack-stage/"*
SPACK_RPATH_DIRS=
SPACK_SHORT_SPEC=test-duplicate@1%gcc@12.2.0 build_system=generic arch=linux-debian12-skylake/yc7izcs
SPACK_STORE_INCLUDE_DIRS=
SPACK_STORE_LINK_DIRS=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/gcc-runtime-12.2.0-u3t2khdtejyh4r7bsmqh67d7vxlq6l7g/lib
SPACK_STORE_RPATH_DIRS=/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/test-duplicate-1-yc7izcsh6n757u43vwzccvocqxospehy/lib:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/test-duplicate-1-yc7izcsh6n757u43vwzccvocqxospehy/lib64:/home/jakob/spack/opt/spack/linux-debian12-skylake/gcc-12.2.0/gcc-runtime-12.2.0-u3t2khdtejyh4r7bsmqh67d7vxlq6l7g/lib
SPACK_SYSTEM_DIRS="/"|"//"|"/bin"|"/bin/"|"/bin64"|"/bin64/"|"/include"|"/include/"|"/lib"|"/lib/"|"/lib64"|"/lib64/"|"/usr"|"/usr/"|"/usr/bin"|"/usr/bin/"|"/usr/bin64"|"/usr/bin64/"|"/usr/include"|"/usr/include/"|"/usr/lib"|"/usr/lib/"|"/usr/lib64"|"/usr/lib64/"|"/usr/local"|"/usr/local/"|"/usr/local/bin"|"/usr/local/bin/"|"/usr/local/bin64"|"/usr/local/bin64/"|"/usr/local/include"|"/usr/local/include/"|"/usr/local/lib"|"/usr/local/lib/"|"/usr/local/lib64"|"/usr/local/lib64/"
SPACK_TARGET_ARGS=-march=skylake -mtune=skylake
```
We would expect just a single version of `py-setuptools` in the `$PYTHONPATH`.

### Information on your system

* **Spack:** 0.23.0.dev0 (b623f58782c53bc465d14b84b44e91ffc2b872cd)
* **Python:** 3.9.12
* **Platform:** linux-debian12-skylake

We use the (default) option `concretizer.duplicate.strategy: minimal`. For `concretizer.duplicate.strategy: full` we see the same behvior. If we chose `concretizer.duplicate.strategy: none`, concretization fails.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by muffgaga**
In #46034, a package X build-depends on a `build-tools`-tagged package, e.g., `py-setuptools`, as well as another package Y that itself depends (build+run) on the same `build-tools` tagged package. In this case, the concretizer puts the possibly two versions of the build tool into separate unification sets, but the run-type dependency requires the
build environment of X to have both versions in its environment — which is illegal.
With this fix, the concretizer puts the two build tool nodes into a new common unification set if there is a build→run-type dependency chain on the build tool ⇒ spack now reports the expected error about not being able to find a common concretization of the build tool.

### CODE REVIEW

**Comment by muffgaga** [comment]
BTW. the fail in `lib/spack/spack/test/concretize.py::TestConcretizeSeparately::test_two_setuptools` looks correct to me (in the sense that the test itself is just wrong)… there are two `py-setuptools` versions that will end up in `py-shapely`'s build environment (I'm trying to get `spack build-env` working for the `duplicates.test` repo but have failed so far :/).

After applying this, we end up with only two build-type versions of `py-setupstools`, which is what should work (and does work):

```diff
diff --git a/var/spack/repos/duplicates.test/packages/py-numpy/package.py b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
index 42cb2aecc7..cfccb64b7e 100644
--- a/var/spack/repos/duplicates.test/packages/py-numpy/package.py
+++ b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
@@ -16,5 +16,5 @@ class PyNumpy(Package):
     version("1.25.0", md5="0123456789abcdef0123456789abcdef")
 
     extends("python")
-    depends_on("py-setuptools@=59", type=("build", "run"))
+    depends_on("py-setuptools@=59", type=("build"))
     depends_on("gmake@4.1", type="build")
```

I think we should also have a inverse test (aka make sure the concretization fails) on the previous state.

**Comment by muffgaga** [comment]
@elenimath suggests this restructuring of the unification sets in order to:

> * keep the `build_generic` unification set containing all the non-`multiple_unification_sets` build dependencies and their run dependencies
> * and for for each parent, we also have an extra unification set, that contains:
>   * all the multiple_unification_sets and their run deps
>   * all the multiple_unification_sets run deps of the non-multiple_unification_sets build deps
    
    
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 18c82474c9..923c195f35 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -95,12 +95,19 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
-unification_set(("build", node(X, Child)), node(X, Child))
+unification_set(("build", ParentNode), node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
      multiple_unification_sets(Child),
      unification_set(SetID, ParentNode).
 
+unification_set(("build", ParentNode), node(X, Child))
+  :- attr("depends_on", ParentNode, NonDupeNode, "build"),
+     not multiple_unification_sets(NonDupeNode),
+     attr("depends_on", NonDupeNode, node(X, Child), "run"),
+     multiple_unification_sets(Child),
+     unification_set(_, ParentNode).
+
 unification_set("generic_build", node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
```
⇒ that looks like the complete solution to me (but I will stop spamming NOW… and wait for comments :) ).

**Comment by haampie** [comment]
Brave to take on this concretizer stuff ;) @alalazo will take a look

**Comment by alalazo** [comment]
@muffgaga I was looking at the issue, and I think this is what I would apply to `develop` as a change:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 18c82474c9..51dbba68c8 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -95,7 +95,7 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
-unification_set(("build", node(X, Child)), node(X, Child))
+unification_set(("build", ParentNode), node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
      multiple_unification_sets(Child),

```
There are two tests failing locally for me, I'll try to check them. 

As you said in https://github.com/spack/spack/pull/46106#issuecomment-2317637177 I think we should have two tests.

1. If it's the root to have a pure build dep on `py-setuptools` and on something that has a run dependency on `py-setuptools`, the `py-setuptools` node should be unique and both edges should point to it (if the problem is solvable)
2. In the opposite case, where the root has the build + run dep on `py-setuptools`  and on something that has only a build dep to py-setuptools, then we should allow for two `py-setuptools` nodes, when required.

**Comment by alalazo** [comment]
@muffgaga ~Go with this change you suggested, thanks https://github.com/spack/spack/pull/46106#issuecomment-2317678517 !~ Possibly a slightly better change:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 18c82474c9..d0ad292beb 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -95,10 +95,9 @@ unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
 unification_set("root", PackageNode) :- attr("root", PackageNode).
 unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
 
-unification_set(("build", node(X, Child)), node(X, Child))
+unification_set(("build", ParentNode), node(X, Child))
   :- attr("depends_on", ParentNode, node(X, Child), Type),
      Type == "build",
-     multiple_unification_sets(Child),
      unification_set(SetID, ParentNode).
 
 unification_set("generic_build", node(X, Child))

```
Rationale is:
1. Each parent gets its own "build environment" where the build deps are unified.
2. Things that don't allow multiple nodes are also thrown into the "generic_build" unification

**Comment by alalazo** [comment]
This diff instead:
```diff
diff --git a/var/spack/repos/duplicates.test/packages/py-numpy/package.py b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
index 42cb2aecc7..e0ca7dfa32 100644
--- a/var/spack/repos/duplicates.test/packages/py-numpy/package.py
+++ b/var/spack/repos/duplicates.test/packages/py-numpy/package.py
@@ -16,5 +16,5 @@ class PyNumpy(Package):
     version("1.25.0", md5="0123456789abcdef0123456789abcdef")
 
     extends("python")
-    depends_on("py-setuptools@=59", type=("build", "run"))
+    depends_on("py-setuptools@=59", type="build")
     depends_on("gmake@4.1", type="build")
diff --git a/var/spack/repos/duplicates.test/packages/py-shapely/package.py b/var/spack/repos/duplicates.test/packages/py-shapely/package.py
index 091f8c9938..f5809d3341 100644
--- a/var/spack/repos/duplicates.test/packages/py-shapely/package.py
+++ b/var/spack/repos/duplicates.test/packages/py-shapely/package.py
@@ -15,4 +15,4 @@ class PyShapely(Package):
 
     extends("python")
     depends_on("py-numpy", type=("build", "link", "run"))
-    depends_on("py-setuptools@=60", type="build")
+    depends_on("py-setuptools@=60", type=("build", "run"))

```
should make the current tests pass. Then we'd just need a test for the case that should fail.

**Comment by muffgaga** [comment]
@alalazo Thanks for looking into this. I'm currently creating another test that checks some "build-tool with its own build deps" case, and I'm not yet sure that everything works there (we don't catch it in the tests, because `py-setuptools` doesn't have deps in those). I also added "fixed" tests (and the inverse one) → it's probably better style to keep the package names the same and add variants for `+rundep` or something like that, right?
(And I'll update the PR according to your suggestions.)

**Comment by elenimath** [comment]
@alalazo about https://github.com/spack/spack/pull/46106#issuecomment-2317678517, I just assumed that smaller unification sets were better performance-wise, but I'm not sure what the actual impact would be :)

**Comment by muffgaga** [comment]
> @alalazo about [#46106 (comment)](https://github.com/spack/spack/pull/46106#issuecomment-2317678517), I just assumed that smaller unification sets were better performance-wise, but I'm not sure what the actual impact would be :)

I did a quick `spack solve --timers -UtI nest` on my laptop (`develop` from this morning) — surely no reproducible state, just to get a rough estimate on the impact:

@alalazo's nice-and-clean patch:
```
    setup          3.699s
    load           0.460s
    ground         5.080s
    solve         10.756s
    total         20.489s
```

@elenimath yours :P
```
    setup          3.757s
    load           0.482s
    ground         4.320s
    solve          6.183s
    total         15.235s
```

**Comment by muffgaga** [comment]
Updated merge branch:

* adapted change in `concretize.lp` (fixes author :))
* duplicates.test repo: PyNumpy has now a `rundeps` variant but defaults to build-only dependency on `py-setuptools`
* duplicates.test repo: PyShapely now run-depends on `py-setuptools` (to check that this is possible)

**Comment by alalazo** [comment]
Looks good, thanks. I'll do some performance benchmark asap, and post it.

**Comment by muffgaga** [comment]
I added another test (to cover "chains" of build tools) that looks like there's a problem with "strategy: full".

The test should cover this case here:
```
#  all build deps
o compiled-tool@1.0
|\
o | seemake@4.0
o | gmake@4.1
 /
o compiled-intermediate-tool@3.0
o seemake@3.0
o gmake@3.0
```

But in `strategy: full` when I look at the `self.{_link_run,_total_build,_direct_build}` members, `gmake` is there only once, i.e. `max_dupes` for it is set to one → fails to concretize.
In `strategy: minimal` it does not happen because `gmake` is a build tool and a max_dupes of 2 is set for it.


I added another test (to cover "chains" of build tools) that shows that there's a problem with "strategy: full".

The test should cover this case here:
```
#  all build deps
o compiled-tool@1.0
|\
o | seemake@4.0
o | gmake@4.1
 /
o compiled-intermediate-tool@3.0
o seemake@3.0
o gmake@3.0
```

But in `strategy: full` when I look at the `self.{_link_run,_total_build,_direct_build}` members, `gmake` is there only once, i.e. `max_dupes` for it is set to one → fails to concretize.
When I look at the members in `FullDuplicatesCounter`, it looks like this:
```
# print(str(list(self._link_run)) + " — " + str(list(self._total_build)) + " — " + str(list(self._direct_build)))
['compiled-tool'] — ['gmake', 'compiled-intermediate-tool', 'seemake'] — ['compiled-intermediate-tool', 'seemake']
```
⇒ i.e. `gmake` only appears once, and `max_dupes` of `gmake` is set to `1`.
In `strategy: minimal` it does not happen because `gmake` is a build tool and a max_dupes of 2 is set for it unconditionally.

`self._total_build` looks like: `{'seemake': {'gmake'}, 'gmake': set(), 'compiled-intermediate-tool': {'seemake'}}` ⇒ and we traverse this as a graph to count the leaves but we only count keys ⇒ I think that's wrong.
@elenimath tried a (cycle-safe) DFS instead of just counting the keys → test works then; I added the "fix" (which is crude, introduces a DFS after we just performed a DFS via `possible_dependencies()`… (cycle check is wrong))) as a commit on top of this PR. However…

⇒ **I think this is an unrelated issue.**
We probably should remove the extra test from there, and create a fresh PR? Yes, no, maybe :) ?

**Comment by muffgaga** [comment]
I removed the "full"-strategy related things from this PR as it is an unrelated issue → @elenimath did a deeper investigation and provided a fresh PR #46169 that covers this separate issue (including the initially failing test and maybe also a fix :) ).

**Comment by alalazo** [comment]
I just benchmarked on:

* **Spack:** 0.23.0.dev0 (1f935ac356b8b7427d72b9226f12bf75e74df120)
* **Python:** 3.11.5
* **Platform:** linux-ubuntu20.04-icelake

[radiuss-pr.csv](https://github.com/user-attachments/files/16844416/radiuss-pr.csv)
[radiuss-develop.csv](https://github.com/user-attachments/files/16844417/radiuss-develop.csv)
[radiuss.txt](https://github.com/user-attachments/files/16844418/radiuss.txt)

radiuss

Using unification sets as "build environments" per node comes with a significant performance hit. I'll check later the patch in https://github.com/spack/spack/pull/46106#issuecomment-2317678517, but it seems to me that this one is overly specific to this case (e.g. why only run dependencies and not link dependencies?). 

I'll try to think how to optimize this use case :thinking:

**Comment by elenimath** [comment]
> e.g. why only run dependencies and not link dependencies?

true, it should be `Type != "build"` instead

**Comment by muffgaga** [comment]
> > e.g. why only run dependencies and not link dependencies?
> 
> true, it should be `Type != "build"` instead

(`"run"` vs. `Type != build` doesn't seem to affect solving time much.)

> Using unification sets as "build environments" per node comes with a significant performance hit. I'll check later the patch in https://github.com/spack/spack/pull/46106#issuecomment-2317678517, […]

Some single `solve --timers -UtI` runs of `glvis` (which shows the hardest impact in your plots) and see for the total times (most of the difference is in the solving time, as your plots also show; `test/concretize.py` (test state of ba77d3) from this does only fail for "develop"); this is all ± 1s depending on what my browser does in parallel ;):

* `develop`: 22.5s
* @alalazo's suggested fix: 38.5s (i.e. +70%)
* @alalazo's [initially suggested fix](https://github.com/spack/spack/pull/46106#issuecomment-2318066147): 28s (i.e. +25%)
* @elenimath's suggested fix (modified, w/ `Type != "build"`): 28s (i.e. +25%)

**Comment by haampie** [comment]
I think it was sensible to leave out type=link. Link deps roughly correspond to local search paths or no search at all, run deps have their global search `*PATH`s, which is the issue being solved.

But of course no reason to not be more restrictive if that improves performance...

File: var/spack/repos/duplicates.test/packages/py-shapely/package.py [hunk 1]
```diff
@@ -15,4 +15,4 @@ class PyShapely(Package):
 
     extends("python")
     depends_on("py-numpy", type=("build", "link", "run"))
-    depends_on("py-setuptools@=60", type="build")
+    depends_on("py-setuptools@=60", type=("build", "run"))
```
Comment (alalazo) [inline var/spack/repos/duplicates.test/packages/py-shapely/package.py]: I think we should avoid this change:
```suggestion
    depends_on("py-setuptools@=60", type="build")
```
Without it, we model the use case in the issue correctly. Besides, with the change, all unit tests pass also on develop (without, instead, 2 tests would fail on develop).

**Review by alalazo — CHANGES_REQUESTED** [review]
I'm currently looking at performance vs. correctness. 

I think the most correct change is the one in https://github.com/spack/spack/pull/46106#issuecomment-2318106245 but it's really slow (due to an explosion of the rules for unification sets).

I'm fine with the heuristic in https://github.com/spack/spack/pull/46106#issuecomment-2317678517. The limitation there is that it doesn't account for transitive dependencies of `NonDupeNode` in the link/run subdag. It's also slower than `develop` due to the same reason as above, but at a lower scale.

TL;DR I'm looking for some alternative heuristic to avoid slowing down the solve. If we can't find it, I'm fine with https://github.com/spack/spack/pull/46106#issuecomment-2317678517. The request for change is for the comment on the test. Can you double check it?
